drivers/dma/dma_silabs_siwx917.c

```diff
@@ -0,0 +1,634 @@
+		irq_number = UDMA0_IRQn;
+	} else {
+		irq_number = UDMA0_IRQn;
+	}
+	// Disable IRQ
+	irq_disable(irq_number);
+	uint8_t channel;
+	uint32_t int_status;
+	uint8_t soft_trig_flag = 0;
+	uint8_t transfer_complete = 0;
+	int_status = cfg->reg->UDMA_DONE_STATUS_REG; // Read the interrupt status
+	// Initialize UDMA Resources
+	UDMA_RESOURCES UDMA_Siwx917_Resources = {
+		cfg->reg, irq_number,
+		(RSI_UDMA_DESC_T *)cfg->reg->CTRL_BASE_PTR // SRAM base address
+	};
+
+	// Identify the interrupt channel
+	for (channel = 0; channel < cfg->channels; channel++) {
+		if (int_status & (1U << channel)) {
+			// Copy channel info to channel_info[].
+			channel_info[channel].Cnt = data[channel].siwx917_udma_channel_info.Cnt;
+			channel_info[channel].DestAddr =
+				data[channel].siwx917_udma_channel_info.DestAddr;
+			channel_info[channel].Size = data[channel].siwx917_udma_channel_info.Size;
+			channel_info[channel].SrcAddr =
+				data[channel].siwx917_udma_channel_info.SrcAddr;
+			if (channel_info[channel].Cnt != channel_info[channel].Size) {
+				// Check if the transfer type is memory-memory
+				if ((UDMA_Siwx917_Resources.desc[channel]
+					     .vsUDMAChaConfigData1.srcInc != UDMA_SRC_INC_NONE) &&
+				    (UDMA_Siwx917_Resources.desc[channel]
+					     .vsUDMAChaConfigData1.dstInc != UDMA_DST_INC_NONE)) {
+					// Need to apply a software trigger later
+					soft_trig_flag = 1;
+				}
+				if ((channel_info[channel].Size - channel_info[channel].Cnt) >
+				    DMA_MAX_TRANSFER_COUNT) {
+					// Update dev->data with latest transfer count.
+					data[channel].siwx917_udma_channel_info.Cnt +=
+						DMA_MAX_TRANSFER_COUNT;
+				} else {
+					data[channel].siwx917_udma_channel_info.Cnt +=
+						(channel_info[channel].Size -
+						 channel_info[channel].Cnt);
+				}
+			} else {
+				transfer_complete = 1;
+			}
+			break;
+		}
+	}
+	// Call UDMA ROM IRQ handler.
+	ROMAPI_UDMA_WRAPPER_API->uDMAx_IRQHandler(&UDMA_Siwx917_Resources,
+						  UDMA_Siwx917_Resources.desc, channel_info);
+
+	if (soft_trig_flag) {
+		// Set the software trigger bit for starting next transfer
+		cfg->reg->CHNL_SW_REQUEST |= (1U << channel);
+	}
+	if ((data[channel].dma_callback != NULL) && transfer_complete) {
+		// Transfer complete, call user callback
+		data[channel].dma_callback(dev, data[channel].cb_data, channel + 1, 0);
+	}
+	// Enable IRQ
+	irq_enable(irq_number);
+}
+
+// Store the Si91x DMA APIs
+static const struct dma_driver_api siwx917_dma_driver_api = {
+	.config = dma_siwx917_configure,
+	.reload = dma_siwx917_reload,
+	.start = dma_siwx917_start,
+	.stop = dma_siwx917_stop,
+	.get_status = dma_siwx917_get_status,
+};
+
+#define SIWX917_DMA_IRQ_CONNECT(inst)                                                              \
+	IRQ_CONNECT(DT_INST_IRQ(inst, irq), DT_INST_IRQ(inst, priority), dma_siwx917_isr,          \
+		    DEVICE_DT_INST_GET(inst), 0);                                                  \
+	irq_enable(DT_INST_IRQ(inst, irq));
```

Comment:
I suggest to inline this macro.

Reply:
Do you mean to use an inline function instead of this macro?